### ISSUE TITLE
Make images more maintainable

### ISSUE THREAD

**Issue by maxolasersquad** [linappleii/linapple#84]
Based on 2b3cf0cf12202f882bdbb703ead044175a17054b

Traditionally the images have been stored in code as both xpm and bmp files. We should store them in a single format that gets "compiled" when make is run.

The splash image should be in a layered format so that the text can be updated fairly easily when needed.

My vote would be to store the images as Gimp xcf files. We should be able to use imagemagic to convert these to bmp when make is run ... I think.

**Comment by ghedger** [comment]
I second this.  I've used Imagemagik in a PHP context, never the C/C++ API.  Worth pursuing; you want to take this or should I?

**Comment by maxolasersquad** [comment]
Here is a stab at removing the forground from the splash image. There are still bits and pieces with the foreground remaining. The Ultima IV screenshot is its own layer here.
[splash.zip](https://github.com/linappleii/linapple/files/4279081/splash.zip)

**Comment by ghedger** [comment]
I can't open the .xcf contained in the zip.  "Unsupported XCF file version 11".  Looks like I need to update my GIMP from 2.8...

Looking at the Magick++ API.  Next challenge will be getting that to the SDL surface.

**Comment by ghedger** [comment]
Also, do we want to consider perhaps updating the background screen (the Apple IIe with Ultima IV) to something else to freshen things up a bit?  The new layer with the version et. al. could then sit on top of this new image.  Just an idea.

**Comment by maxolasersquad** [comment]
I think a new image would be a good idea. The monitor in this image is super weird. Did that aspect ratio exist in the 80s? It looks like a widescreen.

**Comment by ghedger** [comment]
Maybe float something on the Apple II enthusiasts page on FB.  My IIs are all in a box and haven't been fired up in awhile.

**Comment by maxolasersquad** [comment]
I just requested membership. I'll make a post when it gets accepted.

**Comment by ghedger** [comment]
I assumed you were a member.  Anyways, good.  Installing latest GIMP and want to investigate the imagemagick-during-build angle.  That's an elegant solution to this.

**Comment by xandark** [comment]
I'm joining this from the FB post and think it's a great idea to update the splash screen. Maybe something clean and simple, like an old Apple logo on a white background?

![Apple logo](https://user-images.githubusercontent.com/5048913/75746028-cc11ef80-5ccd-11ea-9ea5-237cfbfcbd7b.png)

**Comment by ghedger** [comment]
Okay, I spent a couple hours on this, researching GNU string substitution in Makefiles.  `make` now "builds" the images as the first step in its primary "all" build.  It searches the res/ directory for anything with a `.png` extension (see below) and will convert it to a `.xpm` file.  `sed` then goes in and replaces ${file} with ${file}_xpm so we stick with convention.  This won't solve the 256 color limitation -- will play with that later. 

`convert` behaves flakily with xcf files -- it worked at first but started giving memory allocation errors after awhile.  Using `.png` extensions `convert` seems to operate reliably.

It would be very nice to overlay a "dynamic" image for tips (e.g. "Press CTRL-F2 to reset, F1 for Help" and LinApple v2.2" and overlay this onto the background image.  Some unknowns in this equation are XPM format's ability to handle transparency and SDL's ability to integrate it.  It's probably do-able.

Another thing we don't currently do but should is have the splash.xpm dependent on changes to the splash.png like any other source-target relationship.

I'm not going to merge my branch yet as the splash page is now just the blank one @maxolasersquad cleaned up.

**Comment by ghedger** [comment]
@xandark   I love that logo.  Many fond memories associated with that logo.

Need the more, um, artistically-inclined to help out with this.  To that end, if you'd like to play with it my current work branch is `gregh-image-simplification` -- just convert it to a `png` and replace the existing `res/splash.png`

**Comment by Toolkitman** [comment]
Hello i have paint by mouse a new design for Linapple wit hthe ctrl F2 and F1 help written onscreen i hope you like it! 
![Linapple](https://user-images.githubusercontent.com/23223830/75785709-75a3c000-5d64-11ea-9372-98f77997f348.png)
[Linapple.zip](https://github.com/linappleii/linapple/files/4281784/Linapple.zip)

**Comment by ghedger** [comment]
That's kinda cool.  Love the retro lo-res look.
We're kind of thinking along the lines of having the text be on a separate layer from the background.  That way, if (for example) reset ever becomes something other than CTRL-F2, it's easy to change.
Version # (which should appear on this page) would be easy to change.
Thoughts?

**Comment by Toolkitman** [comment]
I can try sorry for the delay!

Il mer 4 mar 2020, 02:03 Greg Hedger <notifications@github.com> ha scritto:

**Comment by Toolkitman** [comment]
If you can do it let me know! It is ok for me 😁 it is a gift to linapple
by me.

Il mer 4 mar 2020, 04:26 Matteo T. <matteotrev@gmail.com> ha scritto:

**Comment by ghedger** [comment]
Okay, @Toolkitman It's been added.  Thank you!
It's possible we'll take some criticism in that it's too "duplo block" looking, but I think it looks fantastic and captures the essence of the charm of the platform.

The image simplification phase 1 is merged.  I'm not particularly happy with it -- it's a little loop that runs in the Makefile each time a build is run.  It's not particularly obtrusive, but I think it should work more like the relationship between `.cpp` files and `.o` files: if the png is dirty, build the image .xpm files.  I'm not a particularly adept script or Makefile ninja; maybe someone else can have a look at this...

But it works.

**Comment by Toolkitman** [comment]
Ok thank you for your clarifications : ).

Il giorno mer 4 mar 2020 alle ore 05:03 Greg Hedger <
notifications@github.com> ha scritto:

**Comment by Toolkitman** [comment]
I think that if it seems lego it will be happy small and big people! : )

Ciao

Matteo Trevisan

Il giorno mer 4 mar 2020 alle ore 10:43 Matteo T. <matteotrev@gmail.com> ha
scritto:

**Comment by Toolkitman** [comment]
Buddy! I have done my best to help you with the new command, i have built an image without commands. Here it's.
![Linapple](https://user-images.githubusercontent.com/23223830/76134850-e88d8f00-6021-11ea-9046-df4050eea96d.png)
[Linapple.zip](https://github.com/linappleii/linapple/files/4301284/Linapple.zip)

**Comment by ghedger** [comment]
Yep, and it's in the code now.  Pull it down, build it, and run it.

**Comment by Toolkitman** [comment]
Great! I'm sorry but i dont have new Mac os i have an old imac with snow
leopard (i'm not so rich) to try it and i think it dosnt work in it.
I will change the image in Linux.
Let me know if you need more help.


Il giorno sab 7 mar 2020 alle ore 06:40 Greg Hedger <
notifications@github.com> ha scritto:

**Comment by Toolkitman** [comment]
Sorry for my previouse error, it seems from what i have read the Linapple for OS X works from 10.5 OS X Tiger. I will give it a try.
: )

**Comment by ghedger** [comment]
Your latest one (sans "Ctrl-F2 Reset" etc.) has been merged.  There was an error on my part - file needed to be 1) PNG (not XCF -- doesn't seem to convert cleanly with ImageMagick) and 2) named splash.png.  I have made these changes.

`make clean
make
cd bin
./linapple`

to see changes.

Note we do not yet have an overlay.  Any volunteers for this?  I'd like to use the same visual model as the disk chooser but without the dimming _but definitely not_ the implementation approach.  I'd rather focus on refactoring the disk chooser than work on the splash screen.  Plus, the latter is highly-visible for someone who is new to the project and wants their work to have an immediate impact.

Any takers?

**Comment by ThorstenBr** [comment]
Yes, I had noticed the parallel build issue. There's a remaining glitch now, when you remove individual xpm files after an initial build. But there's an easy fix. The make rules should be sane then.

**Comment by rhaleblian** [comment]
Makefile is internally renaming xpm data to have an '_xpm' extension.
This may not be necessary.
IIRC GIMP will name XPMs this way. In the current workflow XPMs are not written from GIMP.
#include s would need to be adjusted. There are only a few.

**Comment by rhaleblian** [comment]
See https://github.com/linappleii/linapple/pull/119 for an alternate way to handle image builds, and frankly some other things as well.

**Comment by rhaleblian** [comment]
Also, i claim that #119 fixes the build race.

**Comment by ghedger** [comment]
I'm traveling and not going to be around for a bit.  Hoping someone else can have a look at your changes, review them, and get them pulled into the repo.  Thank you for your help!

**Comment by rhaleblian** [comment]
Safe travels @ghedger .

#119 is in. Now, the built images get staged to build/obj and in turn are installed. They would also get deployed into a package for package builds by OS maintainers; there's another 2 issues about that.

Note that the above reflects that built images are no longer put back into the source tree.

There is still an 'images' Make rule that let's you build them independently of full builds. Note, though, that since they all need to be compiled into the binary, it's not that useful.

Also, just noting that since all images are built-in, they don't need to be separately deployed to disk. That leave us with just two files -- the configuration template and master floppy -- to deploy. Three files total.

**Comment by rhaleblian** [comment]
This ticket got kinda broad, but nonetheless i'd like to propose it resolved for the most immediate desires. Close?

**Comment by Toolkitman** [comment]
Dear Greag,
I have an issue with linapple, can be a wrong load or something out of my
control, Total replay games compilation files works in Applewin but doesnt
load in linapple can you fix it?
or tell me the procedure to follow?

Thank you
Toolkitman

Il giorno lun 18 mag 2020 alle ore 04:56 Greg Hedger <
notifications@github.com> ha scritto:

**Comment by Nerun** [comment]
@Toolkitman , what font have you used in "Linapple" logo? I have many 8-bit fonts here, and i have tested many "identify fonts from image" sites but didn't find the right one.

**Comment by Toolkitman** [comment]
i dont remember exactly the font i used it's passed too time but it's in
Paint for windows 10.

Il giorno mer 3 feb 2021 alle ore 16:22 Daniel Dias Rodrigues <
notifications@github.com> ha scritto:

**Comment by Toolkitman** [comment]
Select, Copy, Paste the whole words and the game is one : )!

Il giorno mer 3 feb 2021 alle ore 18:51 Matteo T. <matteotrev@gmail.com> ha
scritto:

**Comment by nerun** [comment]
I think we can close this.

**Comment by nerun** [comment]
@maxolasersquad, @rhaleblian, @ghedger, @xandark, @Toolkitman, @ThorstenBr, do you agreed to close this? PR #164 resolved this i think.
